source/exe/win32/service_base.cc

```diff
@@ -0,0 +1,177 @@
+#include <codecvt>
+#include <locale>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+#include "common/event/signal_impl.h"
+
+#include "exe/main_common.h"
+#include "exe/service_base.h"
+
+#include "absl/debugging/symbolize.h"
+
+#define SVCNAME TEXT("ENVOY")
+
+namespace Envoy {
+
+namespace {
+DWORD Win32FromHResult(HRESULT value) { return value & ~0x80070000; }
+
+} // namespace
+
+ServiceBase* ServiceBase::service_static = nullptr;
+
+ServiceBase::ServiceBase(DWORD controlsAccepted) : handle_(0) {
+  status_.dwServiceType = SERVICE_WIN32_OWN_PROCESS;
+  status_.dwCurrentState = SERVICE_START_PENDING;
+  status_.dwControlsAccepted = controlsAccepted;
+}
+
+bool ServiceBase::TryRunAsService(ServiceBase& service) {
+  // We need to be extremely defensive when programming between the start of the program until
+  // `main_common` starts because the loggers have not been initialized
+  //  so we do not have a good way to know what is happening if the program fails.
+  service_static = &service;
+  RELEASE_ASSERT(service_static != nullptr, "Global pointer to service should not be null");
+
+  SERVICE_TABLE_ENTRYA service_table[] = {// Even though the service name is ignored for own process
+                                          // services, it must be a valid string and cannot be 0.
+                                          {SVCNAME, (LPSERVICE_MAIN_FUNCTIONA)ServiceMain},
+                                          // Designates the end of table.
+                                          {0, 0}};
+
+  if (!::StartServiceCtrlDispatcherA(service_table)) {
+    auto last_error = ::GetLastError();
+    if (last_error == ERROR_FAILED_SERVICE_CONTROLLER_CONNECT) {
```

Comment:
Maybe add a comment describing what this error means (that we are running as a console app not a service)

Reply:
Added a bunch of comments in aba15d9ecf2fb289bfbe0e79ecde24531681c7e6 let me know what you think